File: native/machine_linux.c
/* Reference Linux host provider for the neutral machine ABI. */
#define _GNU_SOURCE
#include "dotnet_pal.h"
#include <errno.h>
#include <fcntl.h>
#include <limits.h>
#include <sched.h>
#include <stdlib.h>
#include <sys/resource.h>
#include <sys/sysinfo.h>
#include <unistd.h>
#define CPU_WORDS (DOTNET_PAL_MACHINE_MAX_CPUS / (sizeof(unsigned long) * CHAR_BIT))
static uint32_t number(int selector, uint64_t *out) {
    long v = sysconf(selector);
    if (v < 0) return DOTNET_PAL_UNSUPPORTED;
    *out = (uint64_t)v; return DOTNET_PAL_OK;
}
static uint32_t possible(uint64_t *out) {
    int fd = open("/sys/devices/system/cpu/possible", O_RDONLY | O_CLOEXEC);
    if (fd < 0) return DOTNET_PAL_UNSUPPORTED;
    char text[8193]; size_t used = 0; uint32_t result = DOTNET_PAL_OS_ERROR;
    for (;;) {
        if (used == sizeof(text) - 1) { result = DOTNET_PAL_UNSUPPORTED; break; }
        ssize_t n = read(fd, text + used, sizeof(text) - 1 - used);
        if (n < 0 && errno == EINTR) continue;
        if (n < 0) break;
        if (n == 0) { result = DOTNET_PAL_OK; break; }
        used += (size_t)n;
    }
    close(fd);
    if (result != DOTNET_PAL_OK || !used) return DOTNET_PAL_OS_ERROR;
    text[used] = 0;
    char *p = text; unsigned long previous = 0; int first = 1;
    for (;;) {
        if (*p < '0' || *p > '9') return DOTNET_PAL_OS_ERROR;
        errno = 0; char *end; unsigned long low = strtoul(p, &end, 10), high = low;
        if (errno || end == p) return DOTNET_PAL_OS_ERROR;
        if (*end == '-') {
            p = end + 1; if (*p < '0' || *p > '9') return DOTNET_PAL_OS_ERROR;
            high = strtoul(p, &end, 10);
            if (errno || end == p) return DOTNET_PAL_OS_ERROR;
        }
        if (high < low || high >= DOTNET_PAL_MACHINE_MAX_CPUS || (!first && low <= previous)) return DOTNET_PAL_OS_ERROR;
        previous = high; first = 0;
        if (*end == ',') { p = end + 1; continue; }
        while (*end == '\n' || *end == '\r' || *end == ' ' || *end == '\t') ++end;
        if (*end) return DOTNET_PAL_OS_ERROR;
        *out = previous + 1; return DOTNET_PAL_OK;
    }
}
static uint32_t query(uint32_t kind, uint64_t *out) {
    *out = 0;
    switch (kind) {
    case DOTNET_PAL_MACHINE_ONLINE_CPUS: return number(_SC_NPROCESSORS_ONLN, out);
    case DOTNET_PAL_MACHINE_POSSIBLE_CPUS: return possible(out);
    case DOTNET_PAL_MACHINE_PAGE_BYTES: return number(_SC_PAGESIZE, out);
    case DOTNET_PAL_MACHINE_CACHE_L1: return number(_SC_LEVEL1_DCACHE_SIZE, out);
    case DOTNET_PAL_MACHINE_CACHE_L2: return number(_SC_LEVEL2_CACHE_SIZE, out);
    case DOTNET_PAL_MACHINE_CACHE_L3: return number(_SC_LEVEL3_CACHE_SIZE, out);
    case DOTNET_PAL_MACHINE_CACHE_L4: return number(_SC_LEVEL4_CACHE_SIZE, out);
    case DOTNET_PAL_MACHINE_PHYSICAL_BYTES:
    case DOTNET_PAL_MACHINE_AVAILABLE_BYTES: {
        uint64_t pages, page;
        uint32_t s = number(kind == DOTNET_PAL_MACHINE_PHYSICAL_BYTES ? _SC_PHYS_PAGES : _SC_AVPHYS_PAGES, &pages);
        if (s) return s;
        s = number(_SC_PAGESIZE, &page); if (s) return s;
        if (!page || pages > UINT64_MAX / page) return DOTNET_PAL_OS_ERROR;
        *out = pages * page; return DOTNET_PAL_OK;
    }
    case DOTNET_PAL_MACHINE_ADDRESS_LIMIT: {
        struct rlimit r; if (getrlimit(RLIMIT_AS, &r)) return DOTNET_PAL_OS_ERROR;
        *out = r.rlim_cur == RLIM_INFINITY ? UINT64_MAX : (uint64_t)r.rlim_cur;
        return DOTNET_PAL_OK;
    }
    case DOTNET_PAL_MACHINE_SWAP_BYTES: {
        struct sysinfo s; if (sysinfo(&s) || !s.mem_unit) return DOTNET_PAL_OS_ERROR;
        if ((uint64_t)s.freeswap > UINT64_MAX / s.mem_unit) return DOTNET_PAL_OS_ERROR;
        *out = (uint64_t)s.freeswap * s.mem_unit; return DOTNET_PAL_OK;
    }
    default: return DOTNET_PAL_UNSUPPORTED;
    }
}
static uint32_t affinity(uint32_t *out, size_t capacity, size_t *count) {
    unsigned long bits[CPU_WORDS] = {0}; *count = 0;
    if (sched_getaffinity(getpid(), sizeof(bits), (cpu_set_t *)bits)) return DOTNET_PAL_OS_ERROR;
    for (size_t i = 0; i < DOTNET_PAL_MACHINE_MAX_CPUS; ++i)
        if (bits[i / (sizeof(long)*CHAR_BIT)] & (1ul << (i % (sizeof(long)*CHAR_BIT)))) ++*count;
    if (!*count) return DOTNET_PAL_OS_ERROR;
    if (*count > capacity) return DOTNET_PAL_BUFFER_TOO_SMALL;
    size_t n = 0;
    for (size_t i = 0; i < DOTNET_PAL_MACHINE_MAX_CPUS; ++i)
        if (bits[i / (sizeof(long)*CHAR_BIT)] & (1ul << (i % (sizeof(long)*CHAR_BIT)))) out[n++] = (uint32_t)i;
    return DOTNET_PAL_OK;
}
static uint32_t bind_cpu(uint32_t cpu) {
    if (cpu >= DOTNET_PAL_MACHINE_MAX_CPUS) return DOTNET_PAL_INVALID_ARGUMENT;
    unsigned long bits[CPU_WORDS] = {0};
    bits[cpu / (sizeof(long)*CHAR_BIT)] = 1ul << (cpu % (sizeof(long)*CHAR_BIT));
    return sched_setaffinity(0, sizeof(bits), (cpu_set_t *)bits) ? DOTNET_PAL_OS_ERROR : DOTNET_PAL_OK;
}
static uint32_t current(uint32_t *out) {
    int cpu = sched_getcpu(); *out = UINT32_MAX;
    if (cpu < 0) return DOTNET_PAL_OS_ERROR;
    *out = (uint32_t)cpu; return DOTNET_PAL_OK;
}
const dotnet_pal_host_machine *dotnet_pal_host_machine_v2(void) {
    static const dotnet_pal_host_machine api = {
        {DOTNET_PAL_ABI_VERSION, sizeof(dotnet_pal_host_machine), DOTNET_PAL_CAP_MACHINE},
        {query, affinity, bind_cpu, current, NULL}
    };
    return &api;
}

File: src/machine_linux.rs
//! Linux machine provider. Native layouts stay here, never in the public table.
use crate::machine::*;
use crate::port::{Error, Machine, Result};
use core::mem;
const WORDS: usize = MAX_CPUS / (mem::size_of::<libc::c_ulong>() * 8);
fn scalar(name: i32) -> Result<u64> {
    let n = unsafe { libc::sysconf(name) };
    if n < 0 { Err(Error::Unsupported) } else { Ok(n as u64) }
}
fn bytes(pages: i32) -> Result<u64> {
    scalar(pages)?.checked_mul(scalar(libc::_SC_PAGESIZE)?).ok_or(Error::Os)
}
/// Linux CPU lists may be sparse: a count is not the highest CPU index plus one.
fn possible_list(data: &[u8]) -> Result<u64> {
    let text = core::str::from_utf8(data).map_err(|_| Error::Os)?.trim();
    let mut highest = None;
    for range in text.split(',') {
        let mut parts = range.split('-');
        let low = parts.next().ok_or(Error::Os)?.parse::<u32>().map_err(|_| Error::Os)?;
        let high = match parts.next() { Some(v) => v.parse::<u32>().map_err(|_| Error::Os)?, None => low };
        if parts.next().is_some() || high < low || high as usize >= MAX_CPUS || highest.is_some_and(|last| low <= last) { return Err(Error::Os); }
        highest = Some(high);
    }
    highest.map(|last| last as u64 + 1).ok_or(Error::Os)
}
fn possible() -> Result<u64> {
    let fd = unsafe { libc::open(c"/sys/devices/system/cpu/possible".as_ptr(), libc::O_RDONLY | libc::O_CLOEXEC) };
    if fd < 0 { return Err(Error::Unsupported); }
    let mut buf = [0u8; 8192];
    let mut used = 0;
    let result = loop {
        if used == buf.len() { break Err(Error::Unsupported); }
        let n = unsafe { libc::read(fd, buf.as_mut_ptr().add(used).cast(), buf.len() - used) };
        if n < 0 {
            if unsafe { *libc::__errno_location() } == libc::EINTR { continue; }
            break Err(Error::Os);
        }
        if n == 0 { break possible_list(&buf[..used]); }
        used += n as usize;
    };
    // Never retry close: the descriptor may have been consumed even on EINTR.
    unsafe { libc::close(fd) };
    result
}
impl Machine for crate::linux::Linux {
    fn query(kind: u32) -> Result<u64> {
        match kind {
            ONLINE_CPUS => scalar(libc::_SC_NPROCESSORS_ONLN),
            POSSIBLE_CPUS => possible(),
            PHYSICAL_BYTES => bytes(libc::_SC_PHYS_PAGES),
            AVAILABLE_BYTES => bytes(libc::_SC_AVPHYS_PAGES),
            PAGE_BYTES => scalar(libc::_SC_PAGESIZE),
            CACHE_L1 => scalar(libc::_SC_LEVEL1_DCACHE_SIZE),
            CACHE_L2 => scalar(libc::_SC_LEVEL2_CACHE_SIZE),
            CACHE_L3 => scalar(libc::_SC_LEVEL3_CACHE_SIZE),
            CACHE_L4 => scalar(libc::_SC_LEVEL4_CACHE_SIZE),
            ADDRESS_LIMIT => {
                let mut limit = mem::MaybeUninit::<libc::rlimit>::uninit();
                if unsafe { libc::getrlimit(libc::RLIMIT_AS, limit.as_mut_ptr()) } != 0 { return Err(Error::Os); }
                let limit = unsafe { limit.assume_init() }.rlim_cur;
                Ok(if limit == libc::RLIM_INFINITY { u64::MAX } else { limit as u64 })
            }
            SWAP_BYTES => {
                let mut info = mem::MaybeUninit::<libc::sysinfo>::uninit();
                if unsafe { libc::sysinfo(info.as_mut_ptr()) } != 0 { return Err(Error::Os); }
                let info = unsafe { info.assume_init() };
                (info.freeswap as u64).checked_mul(info.mem_unit as u64).ok_or(Error::Os)
            }
            _ => Err(Error::Unsupported),
        }
    }
    unsafe fn process_affinity(out: *mut u32, capacity: usize) -> Result<CpuList> {
        let mut mask = [0 as libc::c_ulong; WORDS];
        if unsafe { libc::sched_getaffinity(libc::getpid(), mem::size_of_val(&mask), mask.as_mut_ptr().cast()) } != 0 { return Err(Error::Os); }
        let count: usize = mask.iter().map(|v| v.count_ones() as usize).sum();
        if count == 0 { return Err(Error::Os); }
        if count > capacity { return Ok(CpuList::Required(count)); }
        let width = libc::c_ulong::BITS as usize;
        let mut n = 0;
        for (word, &bits) in mask.iter().enumerate() {
            let mut bits = bits;
            while bits != 0 {
                let bit = bits.trailing_zeros() as usize;
                unsafe { out.add(n).write((word * width + bit) as u32) };
                n += 1; bits &= bits - 1;
            }
        }
        Ok(CpuList::Written(n))
    }
    fn bind_current(cpu: u32) -> Result<()> {
        if cpu as usize >= MAX_CPUS { return Err(Error::InvalidArgument); }
        let mut mask = [0 as libc::c_ulong; WORDS];
        let width = libc::c_ulong::BITS as usize;
        mask[cpu as usize / width] = 1 << (cpu as usize % width);
        if unsafe { libc::sched_setaffinity(0, mem::size_of_val(&mask), mask.as_ptr().cast()) } == 0 { Ok(()) } else { Err(Error::Os) }
    }
    fn current_cpu() -> Result<u32> {
        let cpu = unsafe { libc::sched_getcpu() };
        if cpu < 0 { Err(Error::Os) } else { Ok(cpu as u32) }
    }
}
#[cfg(test)]
mod tests {
    use super::*;
    #[test] fn sparse_possible_indices() {
        assert_eq!(possible_list(b"0-3,8,32-63\n"), Ok(64));
        for x in [b"".as_slice(), b"3-2", b"0-2,2", b"0--3", b"1,0", b"65536", b"0-1,", b"max"] { assert!(possible_list(x).is_err(), "{:?}", x); }
    }
}
